File: apple/BeachRamp/BeachStatus/Sources/BeachStatus/Models/Camera.swift
import Foundation

/// A single live beach webcam from the `/api/v2/cameras` roster.
public struct Camera: Codable, Identifiable, Hashable, Sendable {
    public let id: String
    public let name: String
    public let location: String
    public let streamURL: String

    public init(id: String, name: String, location: String, streamURL: String) {
        self.id = id
        self.name = name
        self.location = location
        self.streamURL = streamURL
    }

    enum CodingKeys: String, CodingKey {
        case id
        case name
        case location
        case streamURL = "stream_url"
    }

    /// The stream URL as a `URL`, or `nil` if not yet resolved (empty string,
    /// e.g. a camera the home cron hasn't populated yet).
    public var url: URL? {
        streamURL.isEmpty ? nil : URL(string: streamURL)
    }
}

/// The `/api/v2/cameras` response: the ordered roster plus the default camera id.
public struct CameraRoster: Codable, Sendable {
    public let cameras: [Camera]
    public let defaultID: String

    public init(cameras: [Camera], defaultID: String) {
        self.cameras = cameras
        self.defaultID = defaultID
    }

    enum CodingKeys: String, CodingKey {
        case cameras
        case defaultID = "default_id"
    }

    /// The default camera, or the first in the roster, or nil if empty.
    public var defaultCamera: Camera? {
        cameras.first { $0.id == defaultID } ?? cameras.first
    }
}

File: api/internal/models/camera.go
package models

import "time"

// Camera is a single live beach webcam in the roster. The YouTube live URL is
// stable; HLSURL is the rotating stream URL re-resolved periodically by the
// home yt-dlp cron (residential IP, to avoid datacenter filtering) and pushed
// back via the admin API.
type Camera struct {
	ID         string    `json:"id"`
	Name       string    `json:"name"`
	Location   string    `json:"location"`
	YouTubeURL string    `json:"youtube_url"`
	HLSURL     string    `json:"hls_url"`
	SortOrder  int       `json:"sort_order"`
	UpdatedAt  time.Time `json:"updated_at"`
}
